src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,539 @@
+        let n = params.n as usize;
+
+        for i in 0..n {
+            let prev_idx = (n + i - 1) % n;
+
+            let mut left = z.get_values().clone()[i];
+            let permuted_input_value = &permuted.permuted_input_value.get_values()[i];
+
+            let permuted_table_value = &permuted.permuted_table_value.get_values()[i];
+
+            left *= &(beta + permuted_input_value);
+            left *= &(gamma + permuted_table_value);
+
+            let mut right = z.get_values().clone()[prev_idx];
+            let mut input_term = unpermuted_input_values
+                .iter()
+                .fold(C::Scalar::zero(), |acc, input| {
+                    acc * &theta + &input.get_values()[i]
+                });
+
+            let mut table_term = unpermuted_table_values
+                .iter()
+                .fold(C::Scalar::zero(), |acc, table| {
+                    acc * &theta + &table.get_values()[i]
+                });
+
+            input_term += &beta;
+            table_term += &gamma;
+            right *= &(input_term * &table_term);
+
+            assert_eq!(left, right);
+        }
+
+        let product_blind = Blind(C::Scalar::random());
+        let product_commitment = params.commit_lagrange(&z, product_blind).to_affine();
+        let z = pk.vk.domain.lagrange_to_coeff(z);
+        let product_coset = pk
+            .vk
+            .domain
+            .coeff_to_extended(z.clone(), Rotation::default());
+        let product_inv_coset = pk.vk.domain.coeff_to_extended(z.clone(), Rotation(-1));
+
+        let product = Product::<C> {
+            product_poly: z.clone(),
+            product_coset,
+            product_inv_coset,
+            product_commitment,
+            product_blind,
+        };
+
+        self.product = Some(product.clone());
+        product
+    }
+
+    pub fn construct_constraints(
+        &self,
+        pk: &ProvingKey<C>,
+        beta: C::Scalar,
+        gamma: C::Scalar,
+        theta: C::Scalar,
+        advice_cosets: &[Polynomial<C::Scalar, ExtendedLagrangeCoeff>],
+        fixed_cosets: &[Polynomial<C::Scalar, ExtendedLagrangeCoeff>],
+    ) -> Vec<Polynomial<C::Scalar, ExtendedLagrangeCoeff>> {
+        let permuted = self.permuted.clone().unwrap();
+        let product = self.product.clone().unwrap();
+        let unpermuted_input_cosets: Vec<Polynomial<C::Scalar, ExtendedLagrangeCoeff>> = self
+            .lookup
+            .input_columns
+            .iter()
+            .map(|&input| match input.column_type {
+                Any::Advice => advice_cosets[pk
+                    .vk
+                    .cs
+                    .get_advice_query_index(Column::<Advice>::from(input), 0)]
+                .clone(),
+                Any::Fixed => fixed_cosets[pk
+                    .vk
+                    .cs
+                    .get_fixed_query_index(Column::<Fixed>::from(input), 0)]
+                .clone(),
+                _ => unreachable!(),
```

Comment:
Don't forget to change this as well (and below) if aux columns are allowed in lookups.